cargo-near/src/commands/build_command/docker.rs

```diff
@@ -0,0 +1,226 @@
+                        let mut contract_path = cloned_repo.contract_path.clone();
+                        contract_path.push("contract.wasm");
+                        contract_path
+                    };
+                    std::fs::copy::<std::path::PathBuf, camino::Utf8PathBuf>(
+                        entry.path(),
+                        wasm_path.clone(),
+                    )?;
+
+                    // TODO: ensure fresh
+                    return Ok(util::CompilationArtifact {
+                        path: wasm_path,
+                        fresh: true,
+                        from_docker: true,
+                    });
+                }
+            }
+
+            Err(color_eyre::eyre::eyre!(
+                "Wasm file not found in directory: `{:?}`.",
+                cloned_repo.tmp_contract_path
+            ))
+        } else {
+            println!();
+            println!(
+                "{}",
+                format!(
+                    "See output above ↑↑↑.\nSourceScan command `{:?}` failed with exit status: {status}.",
+                    docker_cmd
+                ).yellow()
+            );
+            println!(
+                "{}",
+                "You can choose to opt out into non-docker build behaviour by using `--no-docker` flag.".cyan()
+            );
+
+            Err(color_eyre::eyre::eyre!(
+                "Reproducible build in docker container failed"
+            ))
+        }
+    }
+
+    fn docker_subprocess_step(
+        self,
+        docker_build_meta: metadata::ReproducibleBuild,
+        cloned_repo: &cloned_repo::ClonedRepo,
+    ) -> color_eyre::eyre::Result<(ExitStatus, Command)> {
+        let mut cargo_args = vec![];
+
+        if self.no_abi {
+            cargo_args.push("--no-abi")
+        }
+        if self.no_embed_abi {
+            cargo_args.push("--no-embed-abi")
+        }
+        if self.no_doc {
+            cargo_args.push("--no-doc")
+        }
+
+        let color = self
+            .color
+            .clone()
+            .unwrap_or(crate::common::ColorPreference::Auto)
+            .to_string();
+        cargo_args.extend(&["--color", &color]);
+        // Cross-platform process ID and timestamp
+        let pid = id().to_string();
+        let timestamp = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .as_secs()
+            .to_string();
+
+        let volume = format!(
+            "{}:/host",
+            cloned_repo
+                .tmp_repo
+                .workdir()
+                .wrap_err("Could not get the working directory for the repository")?
+                .to_string_lossy()
+        );
+        let docker_container_name = format!("cargo-near-{}-{}", timestamp, pid);
+        let docker_image = docker_build_meta.concat_image();
+        let near_build_env_ref = format!("NEAR_BUILD_ENVIRONMENT_REF={}", docker_image);
+
+        // Platform-specific UID/GID retrieval
```

Comment:
This is a pretty pointless comment. More useful comment would be to explain that we only really need to match uid/gid on Linux since Docker is running natively and mounted FS needs to have proper permissions.